Filter excessive attrs from ORCA's DML delete node

InnerLife0 opened the pull request:
Currently, ORCA passes all attributes to DML node regardless of the type of
operation. Passing all user-defined attrs to delete operation causes
unnecessary scans of these attrs. This may lead to performance decreasing,
especially for column-oriented tables. In the most of cases, such attrs may be
omitted. The only exception is the "after" trigger, which needs all attrs to
work.
The goal of this patch is to change the behavior of delete operation and remove
not used attrs from DML node's targetlist. To achieve that, pdrgpcrSource,
stored in CPhysicalDML, is cleared if possible. To use it, execution part of DML
node was also changed. From now, it can work with subset of attrs in target
list, which is close to how postgres works (ModifyTable in [7x](https://github.com/greenplum-db/gpdb/blob/fb5eb8dd6afb1f0acfc0c6a29514a8f671ea2e23/src/backend/executor/nodeModifyTable.c#L3261), [6x](https://github.com/greenplum-db/gpdb/blob/f451b4d1fe84a6d45aaa4d61ffb35c9b14831fa8/src/backend/executor/nodeModifyTable.c#L2398)).
It was decided that ORCA has enough tests for DELETE operation, including one
with triggers. But there is one uncovered case with after trigger on table with
dropped column. The new mdp test file was added for this case. It's suitable for
ORCA unit tests, but a real execution of a script inside gives an error. This
bug is not related to our patch. New ticket was created internally to fix it.
________
Another solution was to separate source and output arrays and pass them to [LogicalDML](https://github.com/greenplum-db/gpdb/blob/f451b4d1fe84a6d45aaa4d61ffb35c9b14831fa8/src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp#L1359) node and [PhysicalDML](https://github.com/greenplum-db/gpdb/blob/f451b4d1fe84a6d45aaa4d61ffb35c9b14831fa8/src/backend/gporca/libgpopt/src/xforms/CXformImplementDML.cpp#L101) node. I preferred as small and as simple solution as possible. It seems that it not adds inconsistency to code and architecture.
________
I may call a historical the reason that ORCA uses all attrs for delete. The code of DML operations is unified and thus, simpler.
Another reason is that delete operations are not a typical workload for GP.
Anyway, the performance of delete queries for ORCA may be changed dramatically.

Here is the example of improved performance.
```
create table test_col(
i1 int, i2 int, i3 int, i4 int,
i5 int, i6 int, i7 int, i8 int,
i9 int, i10 int, i11 int, i12 int,
i13 int, i14 int, i15 int, i16 int
) with (appendonly=true, orientation=column);

insert into test_col select i,i,i,i,i,i,i,i,i,i,i,i,i,i,i,i from generate_series(1,3000000) i;
explain analyze verbose delete from test_col;
```
Before patch:
```
                                                                                                                                    QUERY PLAN                                                                    
                                                                 
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
-----------------------------------------------------------------
 Delete  (cost=0.00..305377.04 rows=1000000 width=1) (actual time=10510.392..14877.302 rows=1001733 loops=1)
   Output: test_col.i1, test_col.i2, test_col.i3, test_col.i4, test_col.i5, test_col.i6, test_col.i7, test_col.i8, test_col.i9, test_col.i10, test_col.i11, test_col.i12, test_col.i13, test_col.i14, test_col.i15,
 test_col.i16, "outer".ColRef_0019, test_col.ctid
   Executor Memory: 1kB  Segments: 3  Max: 1kB (segment 0)
   ->  Result  (cost=0.00..689.54 rows=1000000 width=78) (actual time=10510.183..13969.730 rows=1001733 loops=1)
         Output: test_col.i1, test_col.i2, test_col.i3, test_col.i4, test_col.i5, test_col.i6, test_col.i7, test_col.i8, test_col.i9, test_col.i10, test_col.i11, test_col.i12, test_col.i13, test_col.i14, test_co
l.i15, test_col.i16, test_col.ctid, test_col.gp_segment_id, 0
         ->  Seq Scan on public.test_col  (cost=0.00..473.90 rows=1000000 width=74) (actual time=10510.180..13387.157 rows=1001733 loops=1)
               Output: test_col.i1, test_col.i2, test_col.i3, test_col.i4, test_col.i5, test_col.i6, test_col.i7, test_col.i8, test_col.i9, test_col.i10, test_col.i11, test_col.i12, test_col.i13, test_col.i14, t
est_col.i15, test_col.i16, test_col.ctid, test_col.gp_segment_id
 Planning time: 16.792 ms
   (slice0)    Executor memory: 1786K bytes avg x 3 workers, 1786K bytes max (seg0).
 Memory used:  128000kB
 Optimizer: Pivotal Optimizer (GPORCA)
 Execution time: 18862.796 ms
(12 rows)
```
After patch:
```
                                                              QUERY PLAN                                                              
--------------------------------------------------------------------------------------------------------------------------------------
 Delete  (cost=0.00..55194.00 rows=1000000 width=1) (actual time=0.206..1811.698 rows=1001733 loops=1)
   Output: "outer".ColRef_0019, test_col.ctid
   Executor Memory: 1kB  Segments: 3  Max: 1kB (segment 0)
   ->  Result  (cost=0.00..506.50 rows=1000000 width=14) (actual time=0.099..1264.342 rows=1001733 loops=1)
         Output: test_col.ctid, test_col.gp_segment_id, 0
         ->  Seq Scan on public.test_col  (cost=0.00..473.90 rows=1000000 width=10) (actual time=0.098..891.939 rows=1001733 loops=1)
               Output: test_col.ctid, test_col.gp_segment_id
 Planning time: 58.020 ms
   (slice0)    Executor memory: 265K bytes avg x 3 workers, 265K bytes max (seg0).
 Memory used:  128000kB
 Optimizer: Pivotal Optimizer (GPORCA)
 Execution time: 3003.620 ms
(12 rows)
```
The real performance depends on storage, number of cols and other obvious things. The test above is from my poor storage.
_____
The final commit message should be copied from last commit.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/65027

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1089839

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1089833

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1089840

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1089834

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/65354

BenderArenadata commented:
Failed job ORCA code linter on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1103213

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1103211

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1103205

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1103212

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1103206

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/65429

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1106345

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1106347

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1106338

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1106339

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1106342

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/65481

BenderArenadata commented:
Failed job ORCA code linter on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1108596

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1108594

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1108595

Commit:
Filter excessive attrs from ORCA's DML delete node

Currently, ORCA passes all attributes to DML node regardless of the type of
operation. Passing all user-defined attrs to delete operation causes
unnecessary scans of these attrs. This may lead to performance decreasing,
especially for column-oriented tables. In the most of cases, such attrs may be
omitted. The only exception is the "after" trigger, which needs all attrs to
work.
The goal of this patch is to change the behavior of delete operation and use
empty source array for DML node if possible. To use it, execution part of DML
node was also changed. From now, it can work with subset of attrs in target
list, which is close to how postgres works.
We agreed that ORCA has enough tests for DELETE operation, including one with
triggers. But there is an interesting case with after trigger on table with
dropped column. The new mdp test file was added for this case. It's suitable for
ORCA unit tests, but a real execution of a script inside gives an error. This
bug is not related to our patch; separate ticket was created.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/65581

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1112480

BenderArenadata commented:
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1112479

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1112481

bimboterminator1 commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -4177,7 +4177,11 @@ CTranslatorDXLToPlStmt::TranslateDXLDml(
 							 child_contexts, output_context);
 
-	if (md_rel->HasDroppedColumns())
+	// Create target list with nulls if rel has dropped cols. DELETE may has
```
may have

InnerLife0 replied:
Thanks

bimboterminator1 commented on src/backend/executor/nodeDML.c:
```diff
@@ -281,17 +281,29 @@ ExecInitDML(DML *node, EState *estate, int eflags)
+	 * may cause execution errors.
 	 */
-	TupleDesc cleanTupType = CreateTupleDescCopy(dmlstate->ps.state->es_result_relation_info->ri_RelationDesc->rd_att);
+	if (estate->es_plannedstmt->commandType != CMD_DELETE)
```
How do we take into account the AFTER trigger? Why we don't ignore this condition for that case? You want to say that all the actions below are unnecessary for the case of full target list? If so, why they are necessary for insert /update cases?

InnerLife0 replied:
In my mind, and this is described by a comment, the following "hack" is about reconstructMatchingTupleSlot() used by INSERT and UPDATE only, not by DELETE. So it doesn't matter how much attrs DELETE's targetlist contains here.

bimboterminator1 commented on src/backend/executor/nodeDML.c:
```diff
@@ -281,17 +281,29 @@ ExecInitDML(DML *node, EState *estate, int eflags)
+	 * number of attrs after replacing. ExecDelete() doesn't have such
+	 * machinery, and more, can work with subset of table attrs. So, the code
+	 * below can add attrs, which doesn't exist in original targetlist. This
+	 * may cause execution errors.
```
I propose to rewrite the comment:

The code below is related only to INSERT/UPDATE operations, since they use the full set of relation attributes during the execution.  DELETE operation doesn't require all the attributes, what is represented by a shorter targetList.

And probably add something about the case with after trigger.

InnerLife0 replied:
But my idea was to link this comment with another comment, which explains the code below by itself. I chose this to not rewrite original comment. So, I use ExecInsert() and ExecUpdate() (not INSERT/UPDATE), because this was already used in linked comment.

I think we don't need to mention triggers here as it doesn't matter is there any trigger or not, code works the same.

bimboterminator1 commented on src/backend/gporca/data/dxl/minidump/DeleteWithAfterTriggerDroppedCol.mdp:
```diff
@@ -0,0 +1,336 @@
+explain verbose delete from t where i = 1;
+
+delete from t where i = 1;
+
```
Are there any mdp tests for the triggers? If there are, it seems meaningless to write another one.

InnerLife0 replied:
> But there is an interesting case with after trigger on table with
dropped column. The new mdp test file was added for this case. It's suitable for
ORCA unit tests, but a real execution of a script inside gives an error. This
bug is not related to our patch; separate ticket was created.

This is exactly about mentioned test case. The given scenario was uncovered before patch.

InnerLife0 commented:
> "source array" is not obvious expression.

Trying to imagine, but can't see a case when it can be not obvious. The comment is given in context of changed code. The code changes are small and source attrs are used only ones. However, I can change a comment if it's a real problem.

> Too informal

Can't find this too informal. But anyway, if yes, is it a real problem to write here like so?

> Why do we really talk about it.

To highlight that we don't ignore this problem. Probably, this should be a part of PR message only, not a commit message. Commit message may have a link to existed PR, if any.

Commit:
Filter excessive attrs from ORCA's DML delete node

Currently, ORCA passes all attributes to DML node regardless of the type of
operation. Passing all user-defined attrs to delete operation causes
unnecessary scans of these attrs. This may lead to performance decreasing,
especially for column-oriented tables. In the most of cases, such attrs may be
omitted. The only exception is the "after" trigger, which needs all attrs to
work.
The goal of this patch is to change the behavior of delete operation and remove
not used attrs from DML node's targetlist. To achieve that, pdrgpcrSource,
stored in CPhysicalDML, is cleared if possible. To use it, execution part of DML
node was also changed. From now, it can work with subset of attrs in target
list, which is close to how postgres works.
It was decided that ORCA has enough tests for DELETE operation, including one
with triggers. But there is one uncovered case with after trigger on table with
dropped column. The new mdp test file was added for this case. It's suitable for
ORCA unit tests, but a real execution of a script inside gives an error. This
bug is not related to our patch.

InnerLife0 commented:
During the private talk, @bimboterminator1 described his point of view and I included more fixes to final commit.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/66178

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1137517

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1137518

BenderArenadata commented:
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1137509

Stolb27 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4027-v2

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/66573

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1156787

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1156788
